dask/dataframe/groupby.py

```diff
@@ -1,68 +1,191 @@
+    return g.min()
+
+def _max(g):
+    return g.max()
+
+def _count(g):
+    return g.count()
+
+###############################################################
+# nunique
+###############################################################
+
+def _nunique_df_chunk(df, index):
+    # we call set_index here to force a possibly duplicate index
+    # for our reduce step
+    grouped = (df.groupby(index).apply(pd.DataFrame.drop_duplicates))
+    grouped.index = grouped.index.get_level_values(level=0)
+    return grouped
+
+def _nunique_series_chunk(df, index):
+    assert isinstance(df, pd.Series)
+    if isinstance(index, np.ndarray):
+        assert len(index) == len(df)
+        index = pd.Series(index, index=df.index)
+    grouped = pd.concat([df, index], axis=1).drop_duplicates()
+    return grouped
 
 
 class _GroupBy(object):
+    """ Superclass for DataFrameGroupBy and SeriesGroupBy
+
+    Parameters
+    ----------
+
+    obj: DataFrame or Series
+        DataFrame or Series to be grouped
+    index: str, list or Series
+        The key for grouping
+    kwargs: dict
+        Other keywords passed to groupby
+    """
+
+    def __init__(self, df, index=None, slice=None, **kwargs):
+        assert isinstance(df, (DataFrame, Series))
+        self.obj = df
+
+        # grouping key passed via groupby method
+        self.index = index
+
+        # slicing key applied to _GroupBy instance
+        self._slice = slice
+
+        self.kwargs = kwargs
+
+        if isinstance(index, Series) and df.divisions != index.divisions:
+            msg = ("The Series and index of the groupby"
+                   " must have the same divisions.")
+            raise NotImplementedError(msg)
+
+        if self._is_grouped_by_sliced_column(self.obj, index):
+            # check whether given Series is taken from given df and unchanged.
+            # If any operations are performed, _name will be changed to
+            # e.g. "elemwise-xxxx"
+
+            # if group key (index) is a Series sliced from DataFrame,
+            # emulation must be performed as the same.
+            # otherwise, group key is regarded as a separate column
+            self._pd_cache = self.obj._pd.groupby(self.obj._pd[index.name])
+
+        elif isinstance(self.index, Series):
+            self._pd_cache = self.obj._pd.groupby(self.index._pd)
+        else:
+            self._pd_cache = self.obj._pd.groupby(self.index)
+
+    def _is_grouped_by_sliced_column(self, df, index):
+        """
+        Return whether index is a Series sliced from df
+        """
+        if isinstance(df, DataFrame) and isinstance(index, Series):
+            if (index.name is not None and
+                index._name == df._name + '.' + index.name):
+                return True
+        return False
+
+    @property
+    def _pd(self):
+        """
+        Return an empty pd.DataFrameGroupBy / pd.SeriesGroupBy to emulate
+        calculation result.
+        """
+        return self._pd_cache
```

Comment:
Ping @sinhrks. If there's a good reason for this design, then this looks good to me. I'd like to get this in soon before the next release.